**Preamble**

**Colloboration Policy**. The student is to *explicitly identify* his/her collaborators in the assignment. If the student did not work with anyone, he/she should indicate `Collaborators=['none']`. If the student obtains a solution through research (e.g., on the web), acknowledge the source, but *write up the solution in HIS/HER OWN WORDS*. There will be a one mark penalty if a student fails to indicate his/her collaborators.

**There will be NO EXCEPTIONS to this grading policy.**

# Assignment 1 - Alternative Optimum

If you need help on using Jupyter notebooks, click <a href='#help'>here</a>. 

Objectives:

(a) Familiarize with the PuLP syntax and use PuLP to solve a linear programme (LP).

(b) Determine if a LP is *degenerate* from a (PuLP) solution. In addition, compute alternative optimal solutions (if they exist) from a (PuLP) solution.



## Hooper's Store

Alan intends to sell some Abby's magical potions in Hooper's Store on Sesame street.

According to Alan's estimates, the profit per bottle of Potion $i$ ($i=1,2,3,4$) is as follows. Note that negative profits for Potions 2 and 3 mean that Alan incurs losses when selling these potions.

| Potions | 1 | 2 | 3 | 4 |
|---------|---|---|---|---|
|Profit (per bottle) |110|-30|-56|10 |

However, the production of potions must adhere to the following rules:

[A] $2x_2 + 5.2x_3 + 8.8\le 10x_1\le 7x_3 +13$;

[B] $150x_1 +  10x_4      \le 40x_2 +75x_3 + 145$; 

[C] $110x_2 + 286x_3 + 414\le460x_1 +30x_4$.

Here, $x_i$ denote the number of bottles of Potion $i$ that are produced.


**(a) (5 marks)** You decide to use linear programming to solve Alan's problem. Formulate the linear programming problem and solve it using PuLP. 

NOTE: Your solution will **not** be integer. In (b), we will try to find **integer** optimal solutions.

In [57]:
import pulp

model = pulp.LpProblem("Hooper's Store Part_a", pulp.LpMaximize)

x1 = pulp.LpVariable('x1', lowBound=0)
x2 = pulp.LpVariable('x2', lowBound=0)
x3 = pulp.LpVariable('x3', lowBound=0)
x4 = pulp.LpVariable('x4', lowBound=0)

#objective function
model += 110 * x1 - 30 * x2 - 56 * x3 + 10 * x4, "Alan's Profit"

#Constraints

#splitting Rule A into 2 parts
model += 2 * x2 + 5.2 * x3 + 8.8 <= 10 * x1, "Rule A_1"
model += 10 * x1 <= 7 * x3 + 13, "Rule A_2"
model += 150 * x1 + 10 * x4 <= 40 * x2 + 75 * x3 + 145, "Rule B"
model += 110 * x2 + 286 * x3 + 414 <= 460 * x1 + 30 * x4, "Rule C"

print(model)

model.solve()

print("Maximum: {}".format(pulp.value(model.objective)))

print("x1: {}".format(x1.varValue))
print("x2: {}".format(x2.varValue))
print("x3: {}".format(x3.varValue))
print("x4: {}".format(x4.varValue))


Hooper's Store Part_a:
MAXIMIZE
110*x1 + -30*x2 + -56*x3 + 10*x4 + 0
SUBJECT TO
Rule_A_1: - 10 x1 + 2 x2 + 5.2 x3 <= -8.8

Rule_A_2: 10 x1 - 7 x3 <= 13

Rule_B: 150 x1 - 40 x2 - 75 x3 + 10 x4 <= 145

Rule_C: - 460 x1 + 110 x2 + 286 x3 - 30 x4 <= -414

VARIABLES
x1 Continuous
x2 Continuous
x3 Continuous
x4 Continuous

Maximum: 114.0
x1: 1.3
x2: 2.1
x3: 0.0
x4: 3.4


**(b)** Notice that the solution in (a) is not integer. In other words, $x_1$, $x_2$, $x_3$, and $x_4$ are not all integers. Take the following steps to obtain an integer solution. 


**(i) (2 marks)** Introduce slack variables to the linear program in (a). Solve the new linear program using PuLP. 

NOTE: Your solution will **not** be integer. 

In [58]:
import pulp

slackmodel = pulp.LpProblem("Hooper's Store Part_b_1", pulp.LpMaximize)

x1 = pulp.LpVariable('x1', lowBound=0)
x2 = pulp.LpVariable('x2', lowBound=0)
x3 = pulp.LpVariable('x3', lowBound=0)
x4 = pulp.LpVariable('x4', lowBound=0)
x5 = pulp.LpVariable('x5', lowBound=0)
x6 = pulp.LpVariable('x6', lowBound=0)
x7 = pulp.LpVariable('x7', lowBound=0)
x8 = pulp.LpVariable('x8', lowBound=0)

#objective function
slackmodel += 110 * x1 - 30 * x2 - 56 * x3 + 10 * x4, "Alan's Profit"

#Constraints
slackmodel += -10 * x1 + 2 * x2 + 5.2 * x3 + x5 == -8.8, "Rule A_1"
slackmodel += 10 * x1 - 7 * x3 + x6 == 13, "Rule A_2"
slackmodel += 150 * x1 - 40 * x2  - 75 * x3 + 10 * x4 + x7 == 145, "Rule B"
slackmodel += -460 * x1 + 110 * x2 + 286 * x3 - 30 * x4 + x8 == -414, "Rule C"

print(slackmodel)

slackmodel.solve()

print("Maximum: {}".format(pulp.value(slackmodel.objective)))

print("x1: {}".format(x1.varValue))
print("x2: {}".format(x2.varValue))
print("x3: {}".format(x3.varValue))
print("x4: {}".format(x4.varValue))
print("x5: {}".format(x5.varValue))
print("x6: {}".format(x6.varValue))
print("x7: {}".format(x7.varValue))
print("x8: {}".format(x8.varValue))



Hooper's Store Part_b_1:
MAXIMIZE
110*x1 + -30*x2 + -56*x3 + 10*x4 + 0
SUBJECT TO
Rule_A_1: - 10 x1 + 2 x2 + 5.2 x3 + x5 = -8.8

Rule_A_2: 10 x1 - 7 x3 + x6 = 13

Rule_B: 150 x1 - 40 x2 - 75 x3 + 10 x4 + x7 = 145

Rule_C: - 460 x1 + 110 x2 + 286 x3 - 30 x4 + x8 = -414

VARIABLES
x1 Continuous
x2 Continuous
x3 Continuous
x4 Continuous
x5 Continuous
x6 Continuous
x7 Continuous
x8 Continuous

Maximum: 114.0
x1: 1.3
x2: 2.1
x3: 0.0
x4: 3.4
x5: 0.0
x6: 0.0
x7: 0.0
x8: 55.0


**(ii) (1 mark)** Identify the nonbasic and basic variables.

Basic Variables: **x1, x2, x4, x8** 

Nonbasic Variables:  **x3, x5, x6, x7** 

**(iii) (2 marks)** Rewrite the constraints as a matrix equation of the following form.
$$ U \left( \begin{array}{c} Z \\ x_* \\ x_* \\ x_*\\ x_*\end{array}\right)
\quad
V \left( \begin{array}{c} 1 \\ x_\# \\ x_\# \\ x_\#\\ x_\#\end{array}\right)\,.
$$
Here, $x_*$'s are the basic variables, while $x_\#$'s are the nonbasic variables.


**Write your answer to (iii) here.**
<a id='matrix'></a>

$$ 
\left(
\begin{array}{ccccc}
1  & -110 & 30  & -10 & 0\\
0  & -10  & 2   & 0   & 0\\
0  & 10   & 0   & 0   & 0\\
0  & 150  & -40 & 10  & 0\\
0  & -460 & 110 & -30 & 1\\
\end{array}
\right)
\quad
\left(
\begin{array}{c}
Z\\
x1\\
x2\\
x4\\
x8\\
\end{array}
\right)
\quad
\left(
\begin{array}{ccccc}
0    & -56  &  0   & 0  & 0\\
-8.8 & -5.2 &  -1  & 0  & 0\\
13   & 7    & 0    & -1 & 0\\
145  & 75   & 0    & 0  & -1\\
-414 & -286 & 0    & 0  & 0\\
\end{array}
\right)
\quad
\left(
\begin{array}{c}
1\\
x3\\
x5\\
x6\\
x7\\
\end{array}
\right)\,
$$



**(iv) (2 marks)** Using (iii), explain why the LP has alternative optimum. If you need help on using `numpy` for matrix operations, click <a href='#matrix'>here</a>. 

NOTE: There are numerical issues with `Python`. So you may assume values smaller than $10^{-4}$ to be zero.


In [65]:
import numpy as np 

np.set_printoptions(precision = 2, suppress = True)

# Matrix U & V are just copied from section above. 

U = np.matrix([
    [1,   -110,  30,  -10,  0],
    [0,   -10,   2,   0,    0],
    [0,   10,    0,   0,    0],
    [0,   150,   -40, 10,   0],
    [0,   -460,  110, -30,  1]
]
)

V = np.matrix([
    [0,    -56,  0,   0,  0],
    [-8.8, -5.2, -1,  0,  0],
    [13,   7,    0,   -1, 0],
    [145,  75,   0,   0,  -1],
    [-414, -286, 0,   0,  0]
]
)

M = np.linalg.inv(U)*V
print(M)


[[114.    0.   -5.   -1.   -1. ]
 [  1.3   0.7  -0.   -0.1  -0. ]
 [  2.1   0.9  -0.5  -0.5  -0. ]
 [  3.4   0.6  -2.   -0.5  -0.1]
 [ 55.  -45.   -5.   -6.   -3. ]]


**Write your answer to (iv) here.**


**ANSWER: After representing the BV using NBV from the matrix operation, observe the profit equation,**

$Z = 114 + 0 * x3 - 5 * x5 - x6 - x7 $ 

**The coefficient of x3 term is '0'. Having a value of '0' indicates zero profit increase from x3. Degeneracy occurs and that's how we tell the LP has alternative optimum.**


**(v) (2 marks)** Determine the set of alternative optimum. 

**Write your answer to (v) here.**

**ANSWER: Given that x3, x5, x6, x7 are NBVs, let's make x5=x6=x7=0, and based on the matrix computed from section above,**

$x1 = 1.3 + 0.7x3$;

$x2 = 2.1 + 0.9x3$;

$x4 = 3.4 + 0.6x3$;

$x8 = 55 - 45x3$;

**The complete optimal solution set can be expressed as,** 

$$
\left(
\begin{array}{c}
x1, & x2, & x3,  & x4, & x5, & x6, & x7, & x8\\
\end{array}
\right)
=
\left(
\begin{array}{c}
 1.3 +0.7x3,  & 2.1 +0.9x3, & x3,  & 3.4 + 0.6x3, & 0, & 0, & 0, & 55 - 45x3\\
\end{array}
\right)\,
$$
**with x3 from $0 \le x3 \le 11/9$**

**(vi) (1 mark)** Find an **integer** feasible solution whose objective value is optimal.

In [55]:
#In short, the only change that returns integer feasible solution is to add [cat] parameter when declaring 
#Decision varibles

import pulp

slackmodel = pulp.LpProblem("Hooper's Store Part_Integer", pulp.LpMaximize)

x1 = pulp.LpVariable('x1', lowBound=0, cat='Integer')
x2 = pulp.LpVariable('x2', lowBound=0, cat='Integer')
x3 = pulp.LpVariable('x3', lowBound=0, cat='Integer')
x4 = pulp.LpVariable('x4', lowBound=0, cat='Integer')
x5 = pulp.LpVariable('x5', lowBound=0)
x6 = pulp.LpVariable('x6', lowBound=0)
x7 = pulp.LpVariable('x7', lowBound=0)
x8 = pulp.LpVariable('x8', lowBound=0)

#objective function
slackmodel += 110 * x1 - 30 * x2 - 56 * x3 + 10 * x4, "Alan's Profit"

#Constraints
slackmodel += -10 * x1 + 2 * x2 + 5.2 * x3 + x5 == -8.8, "Rule A_1"
slackmodel += 10 * x1 - 7 * x3 + x6 == 13, "Rule A_2"
slackmodel += 150 * x1 - 40 * x2  - 75 * x3 + 10 * x4 + x7 == 145, "Rule B"
slackmodel += -460 * x1 + 110 * x2 + 286 * x3 - 30 * x4 + x8 == -414, "Rule C"

print(slackmodel)

slackmodel.solve()

print("Maximum: {}".format(pulp.value(slackmodel.objective)))

print("x1: {}".format(x1.varValue))
print("x2: {}".format(x2.varValue))
print("x3: {}".format(x3.varValue))
print("x4: {}".format(x4.varValue))
print("x5: {}".format(x5.varValue))
print("x6: {}".format(x6.varValue))
print("x7: {}".format(x7.varValue))
print("x8: {}".format(x8.varValue))


Hooper's Store Part_Integer:
MAXIMIZE
110*x1 + -30*x2 + -56*x3 + 10*x4 + 0
SUBJECT TO
Rule_A_1: - 10 x1 + 2 x2 + 5.2 x3 + x5 = -8.8

Rule_A_2: 10 x1 - 7 x3 + x6 = 13

Rule_B: 150 x1 - 40 x2 - 75 x3 + 10 x4 + x7 = 145

Rule_C: - 460 x1 + 110 x2 + 286 x3 - 30 x4 + x8 = -414

VARIABLES
0 <= x1 Integer
0 <= x2 Integer
0 <= x3 Integer
0 <= x4 Integer
x5 Continuous
x6 Continuous
x7 Continuous
x8 Continuous

Maximum: 114.0
x1: 2.0
x2: 3.0
x3: 1.0
x4: 4.0
x5: 0.0
x6: 0.0
x7: 0.0
x8: 10.0


---